### ISSUE TITLE
Any benchmarking with nonnegative least squares?

### ISSUE THREAD

**Issue by fatihdinc** [HelmchenLabSoftware/Cascade#49]
Hi again!

I just wanted to check whether there has been any benchmarking with Cascade via non-negative least squares? Is there any particular reason to use least-squares specifically?

If you were interested in discussing further offline, I would be interested in suggesting Cascade as a post-processing pipeline to EXTRACT (https://github.com/schnitzer-lab/EXTRACT-public) after we can figure out a way to perform further benchmarking. Thanks!

**Comment by PTRRupprecht** [comment]
Hi Fatih,

I am not sure whether I completely understand your question. "Non-negative least squares" as an evaluating metric would be the sum of all values where the predictions are higher than the ground truth?
We have illustrated our metrics in the supplementary material, figure S3: https://static-content.springer.com/esm/art%3A10.1038%2Fs41593-021-00895-5/MediaObjects/41593_2021_895_MOESM1_ESM.pdf
Maybe you can use this to help me understand your point. Since we have a ground truth, we simply compare our predictions against it, we do not have to do a NNMF for evaluation.
To train the algorithm, we use the mean squared error, maybe this was also a source of confusion?

And yes, I'd be interested in having Cascade as a suggested potential post-processing pipeline for EXTRACT. I'm not using EXTRACT for a project right now, but I have tested it maybe a year ago, I think it has quite some potential and maybe only needs a larger user base.
You can also get in touch with me by email  @(p.t.r.rupprecht+cascade@gmail.com). We could discuss things in a brief zoom meeting. I'll be back and working with reliable internet access during the first week of January. Before that, I would be restricted to emails or discussions via Github.

**Comment by fatihdinc** [comment]
Hi Peter,

Thank you for your kind response! My apologies, I think I was skipping a few steps along the way. Let me rewind a bit. The NNLS I mentioned is not for evaluating Cascade, rather for correctly estimating the DF/F values. Here is a quick summary:

Let us assume that movies follow a linear model $M = S \times T + \epsilon $, where $S$ are the (non-binary) spatial filters that locate the cells, T are time activities, and $\epsilon$ is the noise. In this scenario, we also assume that $S$ are given, either by circling the neurons by hand, which would lead to binary $S$, or through cell extraction algorithms like EXTRACT. Here are some options for obtaining the traces T:

1) ROI averaging: You could average the activity of the movie $M$ weighted by the spatial filters $S$ such that $T = S^T M$. Forgetting about the matrix notation, for a binary filter, this is simply averaging activity within an ROI.

2) Least-squares: You could compute a least-squares estimate $T = (S^T S)^{-1} S^T M$. This would be equivalent to ROI averaging if there is no overlapping neuropil/other ROIs, and if $S$ are not binarized.

3) NNLS: You could solve the problem:  minimize $|| M - S T ||^2_F$ subject to $T \geq 0$ and some other constraints depending on the algorithm (Suite2p and CAIMAN have exponential kernels which I assume were interfering with how Cascade works).

1) has a large bias for trace estimation, 2) has zero bias but very high variance. 3) seems to be the preferred way so far by the existing cell extraction pipelines. EXTRACT is a robust solver, what robust regression was to linear regression, for this problem. You could think of it more broadly like a tool and a theoretical step forward, instead of a specific cell extraction pipeline. 

As for the user base of EXTRACT: We get that question a lot, especially at SFN this year. We do not seek to increase our user base actively (though it has increased exponentially since the SFN), nor is it within our philosophy that cell extraction should be a black box approach with simple GUIs. We have observations of several instances that incorrect estimation of traces $T$ can significantly alter the scientific conclusions, and have real-life effect. This is partly why we do not actively advertise yet, as I believe this process should not be turned into a plug-in/plug-out concept. Rather, we first try to build necessary learning material for the field when it comes to cell extraction. The case of cell extraction is, in my humble opinion, very similar to where machine learning and Pytorch/SKlearn stand. With these libraries, one does not need to know a single thing about linear algebra to import pre-written functions and train/test models. However, doing this without the appropriate pre-knowledge does have real societal impact. We can talk about this more offline, and I am happy to show you some benchmarking we did for EXTRACT.

Where Cascade comes into play in all this: From what I understand, Cascade is evaluated with DF/F traces that are not estimated trough NNLS or any form of non-negativity constrained robust/linear regression. Non-negativity constraint for estimating DF/F traces can alter the noise distribution significantly, although there are some tricks we used in EXTRACT to get around this. We can talk about it later.

My original ask was to benchmark Cascade on DF/F traces that are estimated through NNLS/non-negative robust regression to show that the non-negativity constraint does not interfere with the final predictions. I suspect that correctly estimating DF/F traces will significantly increase the accuracy on the metrics you mentioned.

I will be closing the issue here and we can continue offline if you wish to discuss further. Also, great work with Cascade, congratulations! As Tony alluded to before, I also find that Cascade is a far better substitute to exponential deconvolution. I also find that Cascade works (e.g. passes my visual inspection of qualitative predictions) like a charm in 1p data extracted from EXTRACT, although I admittedly take the outputs as relative firing rates (even for  the 2p case). Thank you for your time, have a wonderful winter break, and please reach out to me at fdinc@(stanford.edu) when you are back and would like to meet!

(Btw, please remind me to talk about in/out-of-distribution errors if I forget in our short Zoom. I believe, though I might be wrong, in one of your main figures, there is a comparison between several out-of-distribution evaluations with an in-distribution one, and not surprisingly the in-distribution scenario outperforms all. If I am not misunderstanding the scenario, I have a suggestion there. )
